python3Packages.pyqt5: fix build for aarch64-darwin

tanshihaj opened the pull request:
###### Description of changes

As disscused in https://github.com/NixOS/nixpkgs/issues/186489 pyqt5 build for aarch64-darwing can be fixed by bumping min macos sdk version to 11.0.

###### Things done

To make full test you will need to cherry-pick this commit: https://github.com/NixOS/nixpkgs/pull/186598/commits/10f06f9ae7a1da748f9a5a72be4d3605511e1f48, and this one: https://github.com/NixOS/nixpkgs/pull/184560/commits/cf8d724787c0f87c2dae8bdc5cebc86d70111574. Without them Qt builds on aarch64 are broken.

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Commit:
python3Packages.pyqt5: fix build for aarch64-darwin

tpwrules commented:
Using the three commits (two in the text plus the one in the PR), I was able to build and run a simple PyQt5 app on my aarch64-darwin machine running macOS 12.5. I arbitrarily chose `pkgs.retext`.

But the ultimate problem is with SIP. I've submitted an inquiry to their mailing list pointing out this issue so hopefully they fix it soon. They acted quickly to write a patch for the aarch64-linux issue I reported in a very similar vein once I finally figured out how the mailing list worked. Once that makes it into a release this new patch can be removed. Could you add a comment to that effect?

tanshihaj commented:
>  I've submitted an inquiry to their mailing list pointing out this issue so hopefully they fix it soon.

tpwrules commented:
The issue with `QtCoremod.sip:23: syntax error` you are seeing is because PyQt5 is no longer compatible with SIP 4. You can patch PyQt to avoid this issue but really the overlay should stop supplying SIP 4.

The mail I submitted is here: https://www.riverbankcomputing.com/pipermail/pyqt/2022-August/044848.html

Maybe just say `# this patch is due to a bug in SIP and can be removed once SIP is upgraded`

veprbl commented on pkgs/development/python-modules/pyqt/pyqt5-pyproject-darwin-min-sdk.patch:
```diff
@@ -0,0 +1,8 @@
+@@ -16,4 +16,5 @@
+
+ [tool.sip.project]
+ verbose = true
```
Managing inter-dependent patches may be cumbersome in the long-term.

tanshihaj replied:
Do you know how to fix it? Sry I'm really bad with all python stuff, I'm more cpp and go guy(

veprbl replied:
Before your change, this was done not in `patches` but in `postPatch`.

tanshihaj replied:
Yes, I know, you think `postPatch` will be better for this case?

veprbl replied:
Yes

tanshihaj commented:
I've changed way to apply `minimum-macos-version = "11.0"` from patches to `postPatch` hooks. Also I've rebased PR to latest master there is no need to apply https://github.com/NixOS/nixpkgs/commit/10f06f9ae7a1da748f9a5a72be4d3605511e1f48 and https://github.com/NixOS/nixpkgs/commit/cf8d724787c0f87c2dae8bdc5cebc86d70111574. Both this commits are in master now.

veprbl commented on pkgs/development/python-modules/pyqt/5.x.nix:
```diff
@@ -37,11 +38,19 @@ buildPythonPackage rec {
+  # Due to bug in SIP .whl name generation we have to bump minimal macos sdk upto 11.0 for
+  # aarch64-darwin. This patch can be removed once SIP will fix it in upstream,
+  # see https://github.com/NixOS/nixpkgs/pull/186612#issuecomment-1214635456.
+  + lib.optionals (stdenv.isDarwin && stdenv.isAarch64) ''
+    minimum-macos-version = "11.0"
+  '' + ''
```
`lib.optionals` is for lists
```suggestion
  + lib.optionalString (stdenv.isDarwin && stdenv.isAarch64) ''
    minimum-macos-version = "11.0"
  '' + ''
```

veprbl reviewed (approved):
Diff LGTM
